frame/transaction-payment/src/lib.rs

```diff
@@ -163,11 +163,33 @@ impl<T: Trait> Module<T> {
 	) -> BalanceOf<T> where
 		T::Call: Dispatchable<Info=DispatchInfo>,
 	{
-		if info.pays_fee == Pays::Yes {
+		Self::compute_fee_raw(len, info.weight, tip, info.pays_fee)
+	}
+
+	/// Compute the fee for the specified weight.
+	///
+	/// This fee is already adjusted by the per block fee adjustment factor and is therefore
+	/// the share that the weight contributes to the overall fee of a transaction.
+	pub fn weight_to_fee_with_adjustment<Balance>(weight: Weight) -> Balance where
+		Balance: UniqueSaturatedFrom<u128>
+	{
+		let fee = UniqueSaturatedInto::<u128>::unique_saturated_into(Self::weight_to_fee(weight));
+		UniqueSaturatedFrom::unique_saturated_from(
+			NextFeeMultiplier::get().saturating_mul_acc_int(fee)
+		)
+	}
```

Comment:
Would be better to tune `saturating_mul_acc_int ` to work with balance generic directly, rather than passing u128 here. The bounds don't look pretty here.

Reply:
Also, why is the function generic itself? We only want to compute this for `BalanceOf<T>`

Reply:
Lookup the only place where it is called from now: The contracts module. At this place we cannot bound `BalanceOf<T>`  any further for Rust related reasons. Just using the largest concrete type was the best thing I could come up with. 

Please note that everything payment related is done in the balane generic directly.